File: src/pp/old_java/Animal.java
/**
 * 
 */
package org.ahahah.pp.client;

import java.util.ArrayList;

/**
 * @author julien Gaffuri
 *
 */
public class Animal {
	final static int PREY=0;
	final static int PREDATOR=1;
	final static double V_MAX = 0.1;

	Land l;
	double x=0,y=0; //position, in pix
	double vx=0,vy=0; //speed, in pix per ms
	int type;
	ArrayList<Animal> predators=new ArrayList<Animal>();
	ArrayList<Animal> preys=new ArrayList<Animal>();

	public Animal(int type, Land l, double x, double y){
		this.type=type;
		this.l=l;

		if(x<0) this.x=0;
		else if(x>l.w) this.x=l.w;
		else this.x=x;

		if(y<0) this.y=0;
		else if(y>l.h) this.y=l.h;
		else this.y=y;

		l.grid.add(this, this.x, this.y);
	}

	public double d(Animal a){
		return Math.hypot((a.x-x), (a.y-y));
	}

	public void move() {
		l.grid.remove(this, this.x, this.y);

		double a=Math.random()*2*Math.PI;
		double r=0.01*Math.random();

		vx+=r*Math.cos(a)*l.timeStepMs;
		if(vx>V_MAX) vx=V_MAX;
		else if(vx<-V_MAX) vx=-V_MAX;
		vy+=r*Math.sin(a)*l.timeStepMs;
		if(vy>V_MAX) vy=V_MAX;
		else if(vy<-V_MAX) vy=-V_MAX;

		x+=vx*l.timeStepMs;
		if(x<0) x=l.w;
		else if(x>l.w) x=0;

		y+=vy*l.timeStepMs;
		if(y<0) y=l.h;
		else if(y>l.h) y=0;

		l.grid.add(this, this.x, this.y);
	}

	public void observe() {
		ArrayList<Animal> as = l.grid.get(x-l.d, y-l.d, x+l.d, y+l.d);

		preys.clear();
		predators.clear();
		for(Animal a:as){
			if(a==this) continue;
			if(d(a)>l.d) continue;
			if(a.type==Animal.PREY)
				preys.add(a);
			else
				predators.add(a);
		}
	}

	public Animal makeChild() {
		//do better: make child around
		return new Animal(type, l, x, y);
	}

}

File: src/pp/Animal.js

class Animal {

    //Animal.PREY = 0;
    //Animal.PREDATOR = 1;

    constructor(type, l, x = 0, y = 0) {
        this.type = type;
        this.l = l;

        //position
        this.x = x < 0 ? 0 : x > l.w ? l.w : x
        this.y = y < 0 ? 0 : y > l.h ? l.h : y

        //speed
        this.vx = 0
        this.vy = 0;

        //? the ones around ?
        this.predators = []
        this.preys = []

        //add to the spatial index
        l.grid.add(this, this.x, this.y);
    }


    d(a) {
        return Math.hypot((a.x - this.x), (a.y - this.y));
    }

    move() {
        const l = this.l
        l.grid.remove(this, this.x, this.y);

        const a = Math.random() * 2 * Math.PI;
        const r = 0.01 * Math.random();

        this.vx += r * Math.cos(a) * l.timeStepMs;
        if (this.vx > l.V_MAX)
            this.vx = l.V_MAX;
        else if (this.vx < -l.V_MAX)
            this.vx = -l.V_MAX;

        this.vy += r * Math.sin(a) * l.timeStepMs;
        if (this.vy > l.V_MAX)
            this.vy = l.V_MAX;
        else if (this.vy < -l.V_MAX)
            this.vy = -l.V_MAX;

        this.x += this.vx * l.timeStepMs;
        if (this.x < 0) this.x = l.w;
        else if (this.x > l.w) this.x = 0;

        this.y += this.vy * l.timeStepMs;
        if (this.y < 0) this.y = l.h;
        else if (this.y > l.h) this.y = 0;

        //TODO use move instead ?
        l.grid.add(this, this.x, this.y);
    }


    observe() {
        const as = [];

        this.preys = [];
        this.predators = [];

        for (let a of as) {
            if (a == this) continue;
            if (this.d(a) > l.d) continue;
            if (a.type == Animal.PREY)
                this.preys.push(a);
            else
                this.predators.push(a);
        }
    }

    makeChild() {
        //do better: make child around
        return new Animal(this.type, this.l, this.x, this.y);
    }

}
